shared/src/getNewMachineText.ts

```diff
@@ -0,0 +1,135 @@
+import * as prettier from "prettier";
+import { MachineConfig } from "xstate";
+import type { MachineParseResult } from "xstate-parser-demo/src/MachineParseResult";
+import { getRawTextFromNode } from "./getRawTextFromNode";
+import { ImplementationsMetadata } from "./types";
+
+const prettierStartRegex = /^([^{]){1,}/;
+const prettierEndRegex = /([^}]){1,}$/;
+
+const UNWRAP_START = `@UNWRAP_START@`;
+const UNWRAP_END = `@UNWRAP_END@`;
+const markAsUnwrap = (str: string) => {
+  return `${UNWRAP_START}${str}${UNWRAP_END}`;
+};
+
+const UNWRAPPER_REGEX = /"@UNWRAP_START@(.{1,})@UNWRAP_END@"/g;
+
+const STATE_KEYS_TO_PRESERVE = [
+  "context",
+  "tsTypes",
+  "schema",
+  "meta",
+  "data",
+  "delimiter",
+  "preserveActionOrder",
+] as const;
+
+export const getNewMachineText = async ({
+  text,
+  fileName,
+  newConfig,
+  machine,
+  implementations,
+}: {
+  text: string;
+  newConfig: MachineConfig<any, any, any>;
+  fileName: string;
+  machine: MachineParseResult;
+  implementations: ImplementationsMetadata;
+}): Promise<string> => {
+  const nodesToPreserve: string[] = STATE_KEYS_TO_PRESERVE.map((nodeKey) => {
+    const node =
+      machine.ast.definition?.[nodeKey]?._valueNode ||
+      machine.ast.definition?.[nodeKey]?.node;
+    return node
+      ? `\n${nodeKey}: ${getRawTextFromNode(text, node!).replace("\n", " ")},`
+      : "";
+  });
+
+  const config: MachineConfig<any, any, any> = {};
+
+  const getKeyStart = (key: keyof MachineConfig<any, any, any>): number => {
+    const start = machine.ast.definition?.[key]?.node?.start;
+    return start ?? 0;
+  };
+
+  Object.keys(newConfig)
+    .sort((a, b) => {
+      return getKeyStart(a as any) - getKeyStart(b as any);
+    })
+    .forEach((key) => {
+      (config as any)[key] = (newConfig as any)[key];
+    });
+
+  const json = JSON.stringify(
+    config,
+    (key, value) => {
+      if (
+        key === "cond" &&
+        implementations?.guards?.[value]?.jsImplementation
+      ) {
+        return markAsUnwrap(
+          implementations?.guards?.[value]?.jsImplementation!,
+        );
+      }
+      if (
+        key === "src" &&
+        implementations?.services?.[value]?.jsImplementation
```

Comment:
I wonder if instead of grabbing those texts first, then wrapping in special delimiters, then unwrapping etc we couldn't just pass the `oldText` here, keep ranges for those implementations and just slice this stuff from the original here.

Reply:
Could you go into more detail?

Reply:
Instead of using `JSON.stringify` we could just traverse the `newConfig` (after sorting its keys) on our own and stringify stuff as we go. If we'd encounter an inline action/whatever then we could look it by the hashed id in the bookkept stuff where the shape of bookkept information would look something like:
```ts
type InlineImplementation = {
  type: 'inline',
  range: {
    start: number;
    end: number;
  }
}
```

Based on this we could just stringify it directly as:
```js
const implementation = implementations[hashedId]
const str = oldText.slice(implementation.range.start, implementation.range.end)
```

Reply:
Let's add this as a followup PR.

Reply:
Summary from call: we're dissatisfied in general with JSON.stringify and want to find a better solution, but not now.